internal/dataplane/kongstate/kongstate.go

```diff
@@ -196,17 +203,30 @@ func (ks *KongState) FillOverrides(logger logr.Logger, s store.Storer) {
 		}
 	}
 
-	// Upstreams
+	ks.FillUpstreamOverrides(s, failuresCollector)
+}
+
+func (ks *KongState) FillUpstreamOverrides(s store.Storer, failuresCollector *failures.ResourceFailuresCollector) {
 	for i := 0; i < len(ks.Upstreams); i++ {
-		kongIngress, err := getKongIngressForServices(s, ks.Upstreams[i].Service.K8sServices)
+		servicesGroup := lo.Values(ks.Upstreams[i].Service.K8sServices)
+		servicesAsObjects := lo.Map(servicesGroup, func(svc *corev1.Service, _ int) client.Object { return svc })
+
+		kongIngress, err := getKongIngressForServices(s, servicesGroup)
 		if err != nil {
-			logger.Error(err, "failed to fetch KongIngress resource for Services",
-				"names", PrettyPrintServiceList(ks.Upstreams[i].Service.K8sServices))
-			continue
+			failuresCollector.PushResourceFailure(err.Error(), servicesAsObjects...)
+		} else {
+			for _, svc := range ks.Upstreams[i].Service.K8sServices {
+				ks.Upstreams[i].override(kongIngress, svc)
+			}
 		}
 
-		for _, svc := range ks.Upstreams[i].Service.K8sServices {
-			ks.Upstreams[i].override(kongIngress, svc)
+		kongUpstreamPolicy, err := GetKongUpstreamPolicyForServices(s, servicesGroup)
+		if err != nil {
+			failuresCollector.PushResourceFailure(err.Error(), servicesAsObjects...)
```

Comment:
could we directly `continue` here?

Reply:
Technically yes, but I think it'd be safer to keep it this way. It would prevent unintentional `continue` if we added more code in this loop that doesn't depend on `KongUpstreamPolicy`. Plus, it's consistent with how we deal with `KongIngress` above to not interrupt `KongUpstreamPolicy` overrides.